core/src/main/scala/kafka/tools/ConsoleProducer.scala

```diff
@@ -362,11 +374,22 @@ object ConsoleProducer {
     }
 
     private def splitHeaders(headers: String): Array[(String, Array[Byte])] = {
-      headerSeparatorPattern.split(headers).map { pair =>
-        (pair.indexOf(headerKeySeparator), ignoreError) match {
+      headersSeparatorPattern.split(headers).map { pair =>
+        (pair.indexOf(headersKeySeparator), ignoreError) match {
           case (-1, false) => throw new KafkaException(s"No header key separator found in pair '$pair' on line number $lineNumber")
           case (-1, true) => (pair, null)
-          case (i, _) => (pair.substring(0, i), pair.substring(i + headerKeySeparator.length).getBytes(StandardCharsets.UTF_8))
+          case (i, _) =>
+            val headerKey = pair.substring(0, i) match {
+              case k if k == nullMarker =>
+                  throw new KafkaException(s"Header keys should not be equal to the null marker '$nullMarker' as they can't be null")
+              case k => k
+            }
+            val headerValue = pair.substring(i + headersKeySeparator.length) match {
+              case v if v == nullMarker =>
+                null
```

Comment:
nit: I would put `null` on the previous line to stay in line with the second case.